File: src/msgpack/pub_msg.cpp
// publisher.cpp
#include <zmq.hpp>
#include <msgpack.hpp>
#include <iostream>
#include <thread>
#include <chrono>

struct Person {
  std::string name;
  int age;

  // Enable msgpack support
  // MSGPACK_DEFINE(...) = array-based encoding
  MSGPACK_DEFINE_MAP(name, age);
};

int main() {
  zmq::context_t ctx(1);
  zmq::socket_t pub(ctx, zmq::socket_type::pub);
  pub.bind("tcp://*:5555");

  // Give subscribers time to connect
  std::this_thread::sleep_for(std::chrono::seconds(1));

  int count = 0;
  while (true) {
    Person p{ "Alice", 30 + count };

    // Serialize to MessagePack buffer
    msgpack::sbuffer sbuf;
    msgpack::pack(sbuf, p);

    // Topic frame
    const std::string topic = "person";

    // Send multipart: topic -> payload
    pub.send(zmq::buffer(topic), zmq::send_flags::sndmore);
    pub.send(zmq::buffer(sbuf.data(), sbuf.size()),
             zmq::send_flags::none);

    std::cout << "Published [" << topic << "] age=" << p.age << "\n";
    count++;
    std::this_thread::sleep_for(std::chrono::seconds(1));
  }
}

File: scripts/msgpack/pub_msg.py
# publisher.py
import time
import zmq
import msgpack
from dataclasses import asdict
from dataclasses import dataclass


@dataclass
class Person:
    name: str
    age: int

def main():
    context = zmq.Context()
    socket = context.socket(zmq.PUB)
    socket.bind("tcp://*:5555")

    # Give subscribers time to connect
    time.sleep(1)

    count = 0
    while True:
        person = Person(name="Alice", age=30 + count)

        # Convert dataclass → dict → msgpack
        payload = msgpack.packb(asdict(person), use_bin_type=True)

        topic = "person"

        # Send multipart message
        socket.send_multipart([
            topic.encode("utf-8"),
            payload
        ])

        print(f"Published [{topic}] {person}")
        count += 1
        time.sleep(1)

if __name__ == "__main__":
    main()
